### PR TITLE
[RDPHOEN-256] Add linux watchdog

### PR DESCRIPTION
**Pull request by sven-foerster**

### CODE REVIEW

File: recipes-core/busybox/busybox/busybox_watchdog.sh [hunk 1]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+                                   
+echo "Starting watchdog ..."                                    
+watchdog -t 10 -T 20 /dev/watchdog
```
Comment (linkjumper) [inline recipes-core/busybox/busybox/busybox_watchdog.sh:4]: I think the default values ​​30s(reset)/60s(reboot) are ok. No arguments (-t, -T) would need to be given for this.

File: recipes-core/busybox/busybox_%.bbappend [hunk 2]
```diff
@@ -2,6 +2,32 @@ SRC_URI += "file://fragment.cfg"
...
+                ${sysconfdir}/rcS.d/S01watchdog \
+                ${sysconfdir}/init.d/reset_bootcount.sh \
+                ${sysconfdir}/rcS.d/S59reset_bootcount \
+               "
+
+
+initd="${D}${sysconfdir}/init.d"
+rcsd="${D}${sysconfdir}/rcS.d"
+
+do_install_append() {
```
Comment (linkjumper) [inline recipes-core/busybox/busybox_%.bbappend:21]: The watchdog feature must not be added for IRMA6 Release 1.

File: recipes-core/busybox/busybox_%.bbappend [hunk 3]
```diff
@@ -2,6 +2,32 @@ SRC_URI += "file://fragment.cfg"
...
+    install -d ${initd}
+    install -d ${rcsd}
+    
+    #busybox's watchdog support
+    install -m 0755 ${WORKDIR}/busybox_watchdog.sh ${initd}/
+    ln -s -r ${initd}/busybox_watchdog.sh ${rcsd}/S01watchdog
+
+    #reset uboot's bootcount on successfull bootup
```
Comment (linkjumper) [inline recipes-core/busybox/busybox_%.bbappend:29]: If I think about it any longer. The `reset_bootcount.sh` should be moved to swupdate (irma6-swuimage) in an other Pull Request. Please don't add reset_bootcount.sh here.

**Review by linkjumper — CHANGES_REQUESTED** [review]

**Review by iris-ersc — CHANGES_REQUESTED** [review]

File: recipes-core/busybox/busybox_%.bbappend (hunk 2 above)
Comment (Jasper-Ben) [reply recipes-core/busybox/busybox_%.bbappend:18]: I would negate that:
` if [ ${IRMA6_RELEASE} -ne 1 ]`

File: recipes-core/busybox/busybox_%.bbappend [hunk 4]
```diff
@@ -2,6 +2,28 @@ SRC_URI += "file://fragment.cfg"
 SRC_URI += "file://arp.cfg"
 SRC_URI += "file://ntpd.cfg"
 SRC_URI += "file://mdev.conf"
+SRC_URI += "${@'file://enable_watchdog.cfg' if d.getVar('IRMA6_RELEASE') == '2' else ''}"
```
[the remark this answers is not in the document]
Comment (sven-foerster) [reply recipes-core/busybox/busybox_%.bbappend:5]: Inhalt Deiner E-Mail:
+                ${sysconfdir}/rcS.d/S59reset_bootcount \

+rcsd="${D}${sysconfdir}/rcS.d"

steht im Widerspruch zu unten:
+                 ${sysconfdir}/rc5.d/S01watchdog \

+ rc5d="${D}${sysconfdir}/rc5.d"

rcS <-> rc5?

Wie kommt das?

**Review by linkjumper — COMMENTED** [review]
I think @Jasper-Ben is right. You should negate the logic with the queries to something like: IRMA6_RELEASE != 1,  because we want this and all other new features by default except for RELEASE 1.

Nevertheless, the current branch works as expected. Tested on imx8mpevk with irma6-maintenance-uuu for Release 2
**[OK]** Test: Kernel Panic -> reboot after 5s
**[OK]** Test: Nowayout -> 60s reboot after 60s
**[OK]** Test: busybox_watchdog.sh is called on startup


Unfortunately there is still something wrong. It seems to me that there are 2 different watchdog devices and triggering /dev/watchdog kills /dev/watchdog0.
```
root@imx8mpevk:~# dmesg | grep watchdog                                         
[    2.407621] imx2-wdt 30280000.watchdog: timeout 60 sec (nowayout=1)          
[   12.299953] watchdog: watchdog0: nowayout prevents watchdog being stopped!   
[   12.306954] watchdog: watchdog0: watchdog did not stop!
```

**Review by linkjumper — APPROVED** [review]

**Review by Eximmius — APPROVED** [review]
Tested on evk, killing wdt restarts the board after ~60 seconds:
```
587 root      3104 S    watchdog /dev/watchdog
root@imx8mpevk:~# kill -9 587
root@imx8mpevk:~# [  127.728730] watchdog: watchdog0: watchdog did not stop!
Mar  9 12:39:17 imx8mpevk user.crit kernel: [  127.728730] watchdog: watchdog0: watchdog did not stop!

root@imx8mpevk:~# 
U-Boot SPL 2020.04-imx_v2020.04_5.4.24_2.1.0+g4979a99482 (May 30 2020 - 06:50:01 +0000)
DDRINFO: start DRAM init
DDRINFO: DRAM rate 4000MTS
...
```

**Review by Jasper-Ben — APPROVED** [review]

**Comment by linkjumper** [comment]
> I think @Jasper-Ben is right. You should negate the logic with the queries to something like: IRMA6_RELEASE != 1, because we want this and all other new features by default except for RELEASE 1.
> 
> Nevertheless, the current branch works as expected. Tested on imx8mpevk with irma6-maintenance-uuu for Release 2 **[OK]** Test: Kernel Panic -> reboot after 5s **[OK]** Test: Nowayout -> 60s reboot after 60s **[OK]** Test: busybox_watchdog.sh is called on startup
> 
> Unfortunately there is still something wrong. It seems to me that there are 2 different watchdog devices and triggering /dev/watchdog kills /dev/watchdog0.
> 
> ```
> root@imx8mpevk:~# dmesg | grep watchdog                                         
> [    2.407621] imx2-wdt 30280000.watchdog: timeout 60 sec (nowayout=1)          
> [   12.299953] watchdog: watchdog0: nowayout prevents watchdog being stopped!   
> [   12.306954] watchdog: watchdog0: watchdog did not stop!
> ```

See https://github.com/mirror/busybox/blob/1_31_stable/miscutils/watchdog.c:
The warning is generated by the linux watchdog driver because busybox's watchdog closes /dev/watchdog before each open. With a newer busybox version, the warning disappears. An upgrade is currently not mandatory. Here is the fix: https://github.com/mirror/busybox/commit/50a37459ff991cab97e7326490d0637ff1106cc8

**Pull request merged**